ls1intum/Artemis
# `Iris`: Add chat button & widget, and a pop-up window

### Checklist
#### General


- [X] Language: We followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [X] We chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Client
- [X] **Important**: We implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [X] We followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [X] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), we specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
#### Changes affecting Programming Exercises

### Motivation and Context
A chat widget is needed to chat with our chatbot(Iris), and a button is needed to open the chat widget.
A pop-up window is needed to request chat permission from the user.

### Description

- Added a chat button on the right bottom of the exercise page.
- Added a chat widget to chat with Iris. It is opened when the user clicks on the chat button.
- Added a pop-up window to get the permission of the user. It is opened when the user clicks on the chat button. If the user gives the permission, it won't be opened next time, instead the chat widget will be opened.

### Steps for Testing

Prerequisites:
- 1 Student
- 1 Programming Exercise

1. Log in to Artemis.
2. Navigate to a course page to see its exercises.
3. Select a Programming Exercise and navigate to its page.
4. Chat button is on the right bottom of the page.
5. Hover over it to see the helping tooltip. ("Start chatting with Iris!")
6. Click on it too see the pop-up window. 
7. Decline to observe that no window is opened.
8. Accept the permission to see that the windows is opened.
9. Check if the chatbot show the last message at the bottom by sending a lot of messages.
10. Observe the pop-up & chatbot visualisation in the dark mode. (Dialog Box will not be updated for Dark Mode because it will be deprecated in the next PR.)
11. Observe the pop-up & chatbot texts in the German language. (The first message will stay in English because it will be changed later.)


#### Performance Review
#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage

### Screenshots
![image](https://github.com/ls1intum/Artemis/assets/56061820/dd1e198e-7454-45de-bb41-c32f2a56aa9d)
![image](https://github.com/ls1intum/Artemis/assets/56061820/6f02d1d4-e6fe-4206-9cd0-0391915bbb73)
